### PR TITLE
Refactors armor into dedicated subtypes

### LINKED ISSUE #8: `getArmor` is too expensive and called too frequently
Every atom with integrity in the game calls `getArmor`, which maps armor to singletons. This includes armor lists, of which nearly 500 exist in typedefs.

### Estimated Cost

`getArmor` is 285ms over 67,671 calls.

### Solutions

Armor has always been a pain point for maintainers since it's very hard to tell why values are the way there are. Ideally there would be no armor lists and instead, you would do `armor = /datum/armor/fireproof` or similar. Anything that needs custom armor could make its own subtype. These could then be used as keys to a singleton. Customizing armor for admins is already not done by editing armor directly and thus there would be no (necessary) difference in UX there.

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

See title.
## Why It's Good For The Game

Code is cleaner, and more readable/intuitive
Technically closes https://github.com/tgstation/dev-cycles-initiative/issues/8
## Changelog
:cl:
refactor: armor, from the ground up basically
/:cl:

### CODE REVIEW

File: code/datums/armor/armor_types.dm [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+/datum/armor/wintercoat_engineer
+	fire = 20
+
+/datum/armor/wintercoat_ce
+	fire = 30
+	acid = 10
+
+/datum/armor/armor_booster
+	melee = 25
+	bullet = 320
```
Comment (CRITAWAKETS) [inline code/datums/armor/armor_types.dm:473]: i dont think this is meant to have 320 armor

**Review by Fikou — CHANGES_REQUESTED** [review]
i dont know if its in your plans already but the armor files are unreadable and need a lot more splitting up (to categories like structures, machines, armor vests, winter coats, boots, modsuits etc)

**Comment by Mothblocks** [comment]
IMO one thing I was going to ask for was for the armor datums to be defined next to the objects themselves. Lowers surface area.

File: code/__DEFINES/combat.dm [hunk 2]
```diff
@@ -19,26 +19,38 @@
...
+#define MELEE "melee"
 /// Involved in checking the likelyhood of applying a wound to a mob.
 #define WOUND "wound"
-/// Involves being eaten
-#define CONSUME "consume"
+
+#define ARMOR_ALL "all_damage_types"
+
+/// Armor values that are used for damage
+#define ARMOR_LIST_DAMAGE list(BIO, BOMB, BULLET, ENERGY, LASER, MELEE)
```
Comment (Mothblocks) [inline code/__DEFINES/combat.dm:47]: List datums should be `(...)` (this and the other two)

File: code/datums/armor/_atom_armor.dm [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/// Get the atom's armor reference
+/atom/proc/GetArmor()
```
Comment (Mothblocks) [inline code/datums/armor/_atom_armor.dm:2]: snake_case

File: code/datums/armor/_atom_armor.dm [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+	return (armor ||= get_armor_by_type(armor_type))
+
+/// Helper to get a specific rating for the atom's armor
+/atom/proc/GetArmorRating(damage_type)
```
Comment (Mothblocks) [inline code/datums/armor/_atom_armor.dm:6]: snake_case

File: code/datums/armor/_atom_armor.dm [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+	var/datum/armor/armor = GetArmor()
+	return armor.get_rating(damage_type)
+
+/// Sets the armor of this atom to the specified armor
+/atom/proc/SetArmor(datum/armor/armor)
```
Comment (Mothblocks) [inline code/datums/armor/_atom_armor.dm:11]: snake_case

File: code/datums/armor/_atom_armor.dm [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+	if(!(src.armor?.type in GLOB.armor_by_type))
+		qdel(src.armor)
+	src.armor = armor
+
+/// Helper to update the atom's armor to a new armor with the specified rating
+/atom/proc/SetArmorRating(damage_type, rating)
```
Comment (Mothblocks) [inline code/datums/armor/_atom_armor.dm:17]: snake_case

File: code/datums/armor/_armor.dm [hunk 7]
```diff
@@ -0,0 +1,221 @@
+GLOBAL_LIST_INIT(armor_by_type, __generate_armor_cache())
+
+/proc/__generate_armor_cache()
+	. = list()
```
Comment (Mothblocks) [inline code/datums/armor/_armor.dm:4]: You don't need this, nothing in your proc errors, but if you want to keep it, `return .` at the end

File: code/datums/armor/_armor.dm [hunk 8]
```diff
@@ -0,0 +1,221 @@
+GLOBAL_LIST_INIT(armor_by_type, __generate_armor_cache())
```
Comment (Mothblocks) [inline code/datums/armor/_armor.dm:1]: If this is used anywhere else, define it

File: code/datums/armor/_armor.dm [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+/proc/__generate_armor_cache()
```
Comment (Mothblocks) [inline code/datums/armor/_armor.dm:3]: Dunders are really ugly, just remove them

File: code/datums/armor/_armor.dm [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+/datum/armor/immune/generate_new_with_modifiers(list/modifiers)
+	return src
+
+/// Generate a brand new armor datum with the multiplier given, if ARMOR_ALL is specified only that modifer is used
+/datum/armor/proc/generate_new_with_multipliers(list/multipliers)
+	var/datum/armor/new_armor = new
+
+	var/all_keys = ARMOR_LIST_ALL()
+	if(ARMOR_ALL in multipliers)
+		var/mult_all = multipliers[ARMOR_ALL]
```
Comment (Mothblocks) [inline code/datums/armor/_armor.dm:94]: `multiplier` etc

File: code/datums/armor/_armor.dm [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+		if(!mult_all)
+			return src
+		for(var/mod in all_keys)
+			new_armor.vars[mod] = vars[mod] * mult_all
+		return new_armor
+
+	for(var/mod in multipliers)
+		if(!(mod in all_keys)) // Did you know that the false block evaluates faster than the true block?
+			stack_trace("Attempt to call generate_new_with_modifiers with illegal modifier '[mod]'! ignoring it")
```
Comment (Mothblocks) [inline code/datums/armor/_armor.dm:103]:
```suggestion
			stack_trace("Attempt to call generate_new_with_modifiers with illegal modifier '[mod]'! Ignoring it")
```

File: code/datums/armor/_armor.dm [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+	if(ARMOR_ALL in multipliers)
+		var/mult_all = multipliers[ARMOR_ALL]
+		if(!mult_all)
+			return src
+		for(var/mod in all_keys)
+			new_armor.vars[mod] = vars[mod] * mult_all
+		return new_armor
+
+	for(var/mod in multipliers)
+		if(!(mod in all_keys)) // Did you know that the false block evaluates faster than the true block?
```
Comment (Mothblocks) [inline code/datums/armor/_armor.dm:102]: Invert it, `if(!x) else` can be hard to interpret and the speed buff here is barely even noticeable in space init, much less here

File: code/datums/armor/_armor.dm [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+		if(!(value in all_keys))
+			stack_trace("Attempt to call generate_new_with_modifiers with illegal modifier '[value]'! ignoring it")
+		else
+			new_armor.vars[value] = values[value]
+	return new_armor
+
+/datum/armor/immune/generate_new_with_specific(list/values)
+	return src
+
+/datum/armor/proc/get_rating(rating)
```
Comment (Mothblocks) [inline code/datums/armor/_armor.dm:134]: Make sure all these procs are documented

Comment (Mothblocks) [reply code/datums/armor/_armor.dm:134]: They are overridden so it is useful

**Review by Mothblocks — CHANGES_REQUESTED** [review]
I'm assuming everything else isn't worth reviewing

**Comment by Mothblocks** [comment]
I think they're all going to get re-requested when undrafted that's shit

**Comment by NamelessFairy** [comment]
Just wanted to ask since I spotted this PR, how will this change modifying armor values ingame using VV?

**Comment by ZephyrTFA** [comment]
> Just wanted to ask since I spotted this PR, how will this change modifying armor values ingame using VV?

moves it to the atom level, other than that it doesn't

**Comment by Mothblocks** [comment]
@NamelessFairy You already should never have been VV'ing the armor directly, that didn't work to begin with because of how weird the armor system was

**Comment by NamelessFairy** [comment]
> @NamelessFairy You already should never have been VV'ing the armor directly, that didn't work to begin with because of how weird the armor system was. There's a modify armor dropdown in VV that you should be using instead

Thank, didn't know about the VV dropdown and that would explain a lot

**Comment by Mothblocks** [comment]
The edit button on armor really should offer to open that first it's a footgun I've seen a lot

**Comment by timothymtorres** [comment]
This PR appears to have spammed a ton of duplicate armor datums to every single object in the game.  #72348 is one example, but after code diving it does this with every armor... yikes.

**Pull request merged**